### PR TITLE
Improve removal of obsolete indexes

### LINKED ISSUE #1955: `FastDiscardOfDeltas` scanning all nodes and properties on delete
**Memgraph version**
v2.16

**Environment**
Ubuntu 22.04

**Describe the bug**
If any node has been deleted in the system, and the transaction is currently the only one running, it will go through the whole dataset and scan properties.

**To Reproduce**
1. Create 10M nodes in Memgraph
2. Create 10 indices on them
3. create 5 nodes independent of those indices, and put an index on them
4. do detach delete on those nodes

i.e.
```
CREATE CONSTRAINT ON (a:TestNode) ASSERT a.Id is UNIQUE;
Create Index on :TestNode;
Create Index on :TestNode(Id);

CREATE (n:TestNode{Id:1});
CREATE (n:TestNode{Id:2});
CREATE (n:TestNode{Id:3});
CREATE (n:TestNode{Id:4});
CREATE (n:TestNode{Id:5});

MATCH (n:TestNode) RETURN n;

MATCH (n:TestNode) DELETE n;
MATCH (n:TestNode) DETACH DELETE n;
```
ensure that you have enough number of nodes and properties in the system so you can mimick this

**Expected behavior**
The nodes should have been deleted fast

**Additional context**

image

```
void InMemoryStorage::InMemoryAccessor::FastDiscardOfDeltas(uint64_t oldest_active_timestamp,
                                                            std::unique_lock<std::mutex> /*gc_guard*/) {
  auto *mem_storage = static_cast<InMemoryStorage *>(storage_);

  // STEP 1 + STEP 2
  std::list<Gid> current_deleted_vertices;
  std::list<Gid> current_deleted_edges;
  GCRapidDeltaCleanup(current_deleted_vertices, current_deleted_edges);

  // STEP 3) skip_list removals
  if (!current_deleted_vertices.empty()) {
    // 3.a) clear from indexes first

THIS PART WILL GO OVER THE WHOLE DATASET
    std::stop_source dummy;
    mem_storage->indices_.RemoveObsoleteEntries(oldest_active_timestamp, dummy.get_token());
    auto *mem_unique_constraints =
        static_cast<InMemoryUniqueConstraints *>(mem_storage->constraints_.unique_constraints_.get());
    mem_unique_constraints->RemoveObsoleteEntries(oldest_active_timestamp, dummy.get_token());

    // 3.b) remove from veretex skip_list
    auto vertex_acc = mem_storage->vertices_.access();
    for (auto gid : current_deleted_vertices) {
      vertex_acc.remove(gid);
    }
  }

  if (!current_deleted_edges.empty()) {
    // 3.c) remove from edge skip_list
    auto edge_acc = mem_storage->edges_.access();
    auto edge_metadata_acc = mem_storage->edges_metadata_.access();
    for (auto gid : current_deleted_edges) {
      edge_acc.remove(gid);
      edge_metadata_acc.remove(gid);
    }
  }
}
```

I would also put the `FastDiscardOfDeltas` in a flag, so we can disable it optionally for users that are experiencing any issues. By default the fast discard would be turned on

### PR DESCRIPTION
**Pull request by DavIvek**
### Description

While discarding deltas and removing vertices from indexes, idea is to not go trough all indexes but just trough the ones we are deleting.

[master < Task] PR
- [x] Provide the full content or a guide for the final git message
    - **[FINAL GIT MESSAGE]**

### CI Testing Labels
Please select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_

### Documentation checklist
- [x] Add the documentation label tag
- [x] Add the bug / feature label tag
- [x] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [x] Write a release note, including added/changed clauses
    - Faster vertex deletion in cases with numerous index entries
- [x] Link the documentation PR here
    - **[Documentation PR link]**
- [x] Tag someone from docs team in the comments

cc @kgolubic 

closes #1955

### CODE REVIEW

**Comment by DavIvek** [comment]
Flamegraph I got after this changes:

**Comment by DavIvek** [comment]
Flamegraph I got after this changes:
Screenshot from 2024-05-16 13-20-58

**Comment by DavIvek** [comment]
An idea to consider, does it make sense to do something similar in GarbageCollect function?

**Comment by Ignition** [comment]
Problem I think:
```
BEGIN; CREATE (n) SET n:L1 REMOVE n:L1 SET n:L2; ROLLBACK;
```

File: src/storage/v2/inmemory/unique_constraints.cpp [hunk 1]
```diff
@@ -488,13 +489,18 @@ std::vector<std::pair<LabelId, std::set<PropertyId>>> InMemoryUniqueConstraints:
...
-void InMemoryUniqueConstraints::RemoveObsoleteEntries(uint64_t oldest_active_start_timestamp, std::stop_token token) {
+void InMemoryUniqueConstraints::RemoveObsoleteEntries(uint64_t oldest_active_start_timestamp, std::stop_token token,
+                                                      const absl::flat_hash_set<LabelId> &labels) {
   auto maybe_stop = utils::ResettableCounter<2048>();
 
   for (auto &[label_props, storage] : constraints_) {
     // before starting constraint, check if stop_requested
     if (token.stop_requested()) return;
 
+    if (labels.find(label_props.first) == labels.end()) {
```
Comment (Ignition) [inline src/storage/v2/inmemory/unique_constraints.cpp:500]: prefer `!labels.contains(label_props.first)`

File: src/storage/v2/inmemory/storage.cpp [hunk 2]
```diff
@@ -986,6 +993,9 @@ void InMemoryStorage::InMemoryAccessor::GCRapidDeltaCleanup(std::list<Gid> &curr
           if (vertex.deleted) {
             DMG_ASSERT(delta.action == Delta::Action::RECREATE_OBJECT);
             current_deleted_vertices.push_back(vertex.gid);
+            for (const auto &label : vertex.labels) {
+              modified_labels.insert(label);
+            }
```
Comment (Ignition) [inline src/storage/v2/inmemory/storage.cpp:998]:
```suggestion
            modified_labels.insert(vertex.labels.cbegin(), vertex.labels.cend());
```

**Review by Ignition — APPROVED** [review]

**Comment by kgolubic** [comment]
@DavIvek is this OK for RN:

Improved the removal process of obsolete indexes. When discarding deltas and removing vertices from indexes, Memgraph now targets only the relevant indexes instead of processing all indexes, resulting in faster vertex deletion in cases with numerous index entries.
[#2043](https://github.com/memgraph/memgraph/pull/2043)

**Pull request merged**